rust/src/measurements/rappor/mod.rs

```diff
@@ -0,0 +1,138 @@
+use crate::core::{Function, Measurement, PrivacyMap};
+use crate::domains::{BitVector, BitVectorDomain};
+use crate::error::Fallible;
+use crate::measures::MaxDivergence;
+use crate::metrics::DiscreteDistance;
+use crate::traits::{samplers::sample_bernoulli_float, InfDiv, InfLn, InfMul, InfSub};
+
+use bitvec::prelude::{bitvec, Lsb0};
+
+/// Make a Measurement that implements Basic RAPPOR
+///
+/// # Citations
+/// * [RAPPOR: Randomized Aggregatable Privacy-Preserving Ordinal Response](https://arxiv.org/abs/1407.6981)
+///
+/// # Arguments
+/// * `f` - Per-bit flipping probability. Must be in $(0, 1]$.
+/// * `m` - number of ones set in each boolean vector (1 if one-hot encoding, more if using a bloom filter)
+///
+/// eps = 2mln((2-f)/f)
+pub fn make_rappor(
+    input_domain: BitVectorDomain,
+    input_metric: DiscreteDistance,
+    f: f64,
+    constant_time: bool,
+) -> Fallible<Measurement<BitVectorDomain, BitVector, DiscreteDistance, MaxDivergence<f64>>> {
+    let m = match input_domain.max_weight {
+        Some(m) => m,
+        None => {
+            return fallible!(
+                MakeMeasurement,
+                "RAPPOR requires a maximum number of set bits!"
+            )
+        }
+    };
+
+    if f <= 0.0 || f > 1.0 {
+        return fallible!(MakeMeasurement, "f must be in (0, 1]");
+    };
+
+    // priv = 2mln((2-f)/f)
+    let epsilon = (2.0f64)
+        .inf_sub(&f)?
+        .inf_div(&f)?
+        .inf_ln()?
+        .inf_mul(&2.0)?
+        .inf_mul(&f64::from(m))?;
+    let f_2 = f.inf_div(&2.0)?;
+    Measurement::new(
+        input_domain,
+        Function::new_fallible(move |arg: &BitVector| {
+            let n = arg.len();
+            let noise_vector = (1..n)
+                .into_iter()
+                .map(|_| sample_bernoulli_float(f_2, constant_time))
+                .collect::<Fallible<BitVector>>()?;
+            // I wanted to avoid cloning here but the closure makes it necessary
+            // Shouldn't use much memory anyway given bit-vecs
+            Ok(arg.clone() ^ noise_vector) // xor on bit vectors
```

Comment:
At some point I'd like to adjust the signature of functions to take data by ownership, instead of by reference. I think this will result in less cloning overall.